File: lib/teien/actors/sphere_actor.rb
require_relative '../actor'

module Teien
  class SphereActor <  Teien::Actor
    attr_accessor :mass
    attr_accessor :col_obj
    attr_accessor :inertia
    attr_accessor :rigid_body
    
    def initialize(name, radius = 1, physics_info = nil, ext_info = nil)
      super(name, physics_info, ext_info)
      @type = "Sphere"
      @radius = radius
    end
    
    def setup(physics)
      # create a physics object.
      @col_obj = Bullet::BtSphereShape.new(@radius)
      @inertia = Bullet::BtVector3.new()
      @col_obj.calculate_local_inertia(@physics_info[:mass], @inertia)
      @rigid_body = Bullet::BtRigidBody.new(@physics_info[:mass], self, @col_obj, inertia)
      super(physics)
      physics.add_rigid_body(@rigid_body)
    end
    
    def to_hash
      hash = super
      hash[:type] = @type
      hash[:radius] = @radius
      hash
    end
    
    def from_hash(hash)
      super(hash)
      @radius = hash[:radius]
    end
  end
end

File: public/teien/actors/sphere_actor.js
SphereActor = function(param){
    teien.Actor.call(this, param);
    this.type = param.type;
    this.radius = param.radius;
};
SphereActor.prototype = Object.create(teien.Actor.prototype);

SphereActor.prototype.setup = function(physics){
    var col_obj = new Ammo.btSphereShape(this.radius);
    var inertia = new Ammo.btVector3();
    col_obj.calculateLocalInertia(this.physics_info.mass, inertia);
    // Ammo does not support btMotionState handling.
    //this.rigidBody = new Ammo.btRigidBody(this.mass, this, col_obj, inertia);
    var ms = new Ammo.btDefaultMotionState(this.transform);
    this.rigidBody = new Ammo.btRigidBody(this.physics_info.mass, ms, col_obj, inertia);
    physics.addRigidBody(this.rigidBody);

    teien.Actor.prototype.setup.call(this, physics);
};

SphereActor.prototype.fromHash = function(param){
    teien.Actor.prototype.fromHash.call(this, param);
    this.radius = param.radius;
};

SphereActor.prototype.update = function(params){
    
};
